lib/scholar/linear/isotonic_regression.ex

```diff
@@ -0,0 +1,419 @@
+          1.0
+        ),
+        x_max: Nx.tensor(
+          11.0
+        ),
+        x_thresholds: Nx.tensor(
+          [1.0, 4.0, 7.0, 9.0, 10.0, 11.0]
+        ),
+        y_thresholds: Nx.tensor(
+          [1.0, 3.0, 6.0, 8.0, 9.0, 10.0]
+        ),
+        increasing?: true,
+        sample_weights: Nx.tensor(
+          [1.0, 1.0, 1.0, 1.0, 1.0, 1.0]
+        ),
+        cutoff_index: Nx.tensor(
+          5
+        )
+      }
+  """
+  deftransform fit(x, y, opts \\ []) do
+    opts = NimbleOptions.validate!(opts, @opts_schema)
+
+    opts =
+      [
+        sample_weights_flag: opts[:sample_weights] != nil
+      ] ++
+        opts
+
+    {sample_weights, opts} = Keyword.pop(opts, :sample_weights, 1.0)
+    x_type = to_float_type(x)
+    x = to_float(x)
+    y = to_float(y)
+
+    sample_weights =
+      if Nx.is_tensor(sample_weights),
+        do: Nx.as_type(sample_weights, x_type),
+        else: Nx.tensor(sample_weights, type: x_type)
+
+    sample_weights = Nx.broadcast(sample_weights, {Nx.axis_size(y, 0)})
+
+    fit_n(x, y, sample_weights, opts)
+  end
+
+  defnp fit_n(x, y, sample_weights, opts) do
+    {x_min, x_max, x_unique, y, index_cut} = build_y(x, y, sample_weights, opts)
+
+    %__MODULE__{
+      x_min: x_min,
+      x_max: x_max,
+      x_thresholds: x_unique,
+      y_thresholds: y,
+      increasing?: opts[:increasing?],
+      cutoff_index: index_cut
+    }
+  end
+
+  @doc """
+  Makes predictions with the given `model` on input `x` and interpolating `function`.
+
+  ## Examples
+
+      iex> x = Nx.tensor([1, 4, 7, 9, 10, 11])
+      iex> y = Nx.tensor([1, 3, 6, 8, 9, 10])
+      iex> model = Scholar.Linear.IsotonicRegression.fit(x, y)
+      iex> function = Scholar.Linear.IsotonicRegression.build_function(model)
+      iex> to_predict = Nx.tensor([1, 2, 3, 4, 5, 6, 7, 8, 9, 10])
+      iex> Scholar.Linear.IsotonicRegression.predict(function, model, to_predict)
+      #Nx.Tensor<
+        f32[10]
+        [1.0, 1.6666667461395264, 2.3333334922790527, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0]
+      >
+  """
+  defn predict(function, model, x) do
+    check_input_shape(x)
+    x = Nx.flatten(x)
+    x = Nx.clip(x, model.x_min, model.x_max)
+    Scholar.Interpolation.Linear.predict(function, x)
+  end
+
+  @doc """
+  Builds a function for predictions with the given `model`.
+
+  ## Examples
+
+      iex> x = Nx.tensor([1, 4, 7, 9, 10, 11])
+      iex> y = Nx.tensor([1, 3, 6, 8, 9, 10])
+      iex> model = Scholar.Linear.IsotonicRegression.fit(x, y)
+      iex> function = Scholar.Linear.IsotonicRegression.build_function(model)
+      %Scholar.Interpolation.Linear{
+        coefficients: Nx.tensor(
+          [
+            [0.6666666865348816, 0.3333333134651184],
+            [1.0, -1.0],
+            [1.0, -1.0],
+            [1.0, -1.0],
+            [1.0, -1.0]
+          ]
+        ),
+        x: Nx.tensor(
+          [1.0, 4.0, 7.0, 9.0, 10.0]
+        )
+      }
+  """
+  def build_function(model) do
```

Comment:
What would this function be used for? The docs says what it does, but not when/why/how we use it. :)

Reply:
This function is essential because based on thresholds that are calculated in isotonic regression it creates a linear interpolation model that generates the results of actual predictions. It is called after fit function when we have all results from isotonic regression.

Reply:
Ah, and if I understand this right, this function cannot be used as `defn` right? We had other models that had some code that could not be executed inside `defn` right? Do you remember which model it was? Maybe we should copy the same structure that we have there?

Reply:
I write it as a `def` because through the whole algorithm, I propagate tensors that normally would have been reduced to smaller tensors (in `make_unique`). So at the very end, I cut them and pass them to `Scholar.Interpolation.Linear`. I think that we have some functions in `Scholar.Metrics.Classification` like `roc_curve` which use `def`.

Reply:
Let me rephrase the question: why don't we do this inside `fit` then?

Reply:
Oh, I found it. it was AffinityPropagation, where we added an optional `prune` function. Could we do something similar here? You can optionally prune but, if you don't, we run with the whole tensor?

Reply:
Hmm, If we pass unpruned tensor to interpolation, then the results would be wrong. The problem with fit and this functionality is that the cutoff index is calculated based on the content of the tensor. So we simply would need a dynamic shape to support this. I would think if we could somehow force users to preprocess data before so we don't need to prune it

Reply:
Okay, I think that we can preprocess the data so we can put this interpolation inside defn, but to be honest this solution is worse than what we have now. Because it will introduce another `def` function that will change the shapes of tensors. 

The solution will be the following: For each data point from x: if there are some indices for which values in `X` are the same, take a weighted (by sample weights) mean of those values, remove all repetitions from X and the according values from `y` and replace them with the calculated mean.

Reply:
Hrm, what do you think about this approach.

We add a new field called `preprocess: nil | Scholar.Interpolation.Linear.t`. You must call `preprocess` before you call `predict` to fill in the preprocess field. In `predict`, if `preprocess` is nil, we raise.

The goal is to keep everything in a single struct. Thoughts?

Reply:
Sure, it will be convenient to have it in one place :)

Reply:
@josevalim I pushed changes. I changed the `preprocess` default value from `nil` to `{}` cause the protocol container is not implemented for `nil`. I added tests. The only thing I need to add is support for decreasing sequences.